lib/Coroutine.php

```diff
@@ -15,126 +15,150 @@ final class Coroutine implements Promise
+                        $yielded = $generator->throw($exception);
                     } else {
                         // Send the new value and execute to next yield statement.
-                        $yielded = $this->generator->send($this->value);
+                        $yielded = $generator->send($value);
                     }
 
                     if (!$yielded instanceof Promise) {
-                        if (!$this->generator->valid()) {
-                            $this->resolve($this->generator->getReturn());
-                            $this->onResolve = null;
+                        if (!$generator->valid()) {
+                            $this->resolve($generator->getReturn());
+                            $onResolve = null;
                             return;
                         }
 
-                        $yielded = $this->transform($yielded);
+                        $yielded = self::transform($yielded, $generator);
                     }
 
-                    $this->immediate = false;
-                    $yielded->onResolve($this->onResolve);
-                } while ($this->immediate);
+                    $immediate = false;
+                    $yielded->onResolve($onResolve);
+                } while ($immediate);
 
-                $this->immediate = true;
+                $immediate = true;
             } catch (\Throwable $exception) {
-                $this->fail($exception);
-                $this->onResolve = null;
+                try {
+                    $this->fail($exception);
+                    $onResolve = null;
+                } catch (\Throwable $e) {
+                    Loop::defer(static function () use ($e) {
+                        throw $e;
+                    });
+                }
             } finally {
-                $this->exception = null;
-                $this->value = null;
+                try {
+                    $exception = null;
+                    $value = null;
+                } catch (\Throwable $e) {
+                    Loop::defer(static function () use ($e) {
+                        throw $e;
+                    });
+                }
             }
         };
 
-        $yielded->onResolve($this->onResolve);
+        try {
+            $yielded->onResolve($onResolve);
+
+            unset($generator, $yielded, $exception, $value, $immediate, $onResolve);
+        } catch (\Throwable $e) {
+            Loop::defer(static function () use ($e) {
+                throw $e;
+            });
+        }
     }
 
-    /**
-     * Attempts to transform the non-promise yielded from the generator into a promise, otherwise returns an instance
-     * `Amp\Failure` failed with an instance of `Amp\InvalidYieldError`.
-     *
-     * @param mixed $yielded Non-promise yielded from generator.
-     *
-     * @return \Amp\Promise
-     */
-    private function transform($yielded): Promise
+    public function __destruct()
```

Comment:
Should this destructor be part of `Placeholder`?

Reply:
Done.